src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -1612,46 +1588,77 @@ private bool DoesParamArgRankMatch(List<Type> parameterTypes, List<AssociativeNo
                             break;
                     }
                 }
-                else if (args[argIndex] is ExprListNode exp)
+                else if (currentArg is ExprListNode exp)
                 {
-                    if (!ArrayUtils.IsEnumerable(p) || p == typeof(string))
-                    {
-                        // arg is an enumerable type, param is not.
-                        return false;
-                    }
-                    var argRank = GetArgumentRank(exp);
-                    if (argRank == -1)
+                    if (exp.ReplicationGuides.Count > 0) return false;
+                    Type t;
+                    if (!astTypeInfoMap.TryGetValue(exp.ID, out t))
                     {
-                        // non-rectangular (jagged) array best handled by replication
-                        return false;
+                        throw new Exception("unkown ast type");
                     }
-                    if (argRank != GetRank(p)) return false;
+                    if (!DoesParamArgRankMatchInner(fep, p, t, argIndex)) return false;
                 }
-                else if (args[argIndex] is IdentifierNode idn)
+                else if (currentArg is IdentifierNode idn)
                 {
                     if (idn.ReplicationGuides.Count > 0) return false;
 
                     var t = variables[idn.Value].Item2;
                     if (t == typeof(object)) return false;
 
-                    if (!p.Equals(t))
+                    if(!DoesParamArgRankMatchInner(fep,p, t,argIndex)) return false;
+                }
+                else if (currentArg is RangeExprNode rgnNode)
+                {
+                    Type t;
+                    if (!astTypeInfoMap.TryGetValue(rgnNode.ID, out t))
                     {
-                        var argRank = GetRank(t);
-                        var paramRank = GetRank(p);
-                        if (argRank != paramRank) return false;
-
-                        // If both have rank 0, it could also be because their type info is ambiguous.
-                        if (argRank == 0 && paramRank == 0) return false;
+                        throw new Exception("unkown ast type");
                     }
+                    if (!DoesParamArgRankMatchInner(fep,p, t,argIndex)) return false;
                 }
                 else return false;
                 argIndex++;
             }
             return true;
         }
 
-        private static int GetRank(Type type)
+        private static bool DoesParamArgRankMatchInner(ProtoCore.CLRFunctionEndPoint fep,Type p, Type t, int paramIndex)
         {
```

Comment:
`DoesParamArgRankMatchInner(ProtoCore.CLRFunctionEndPoint fep,Type paramType, Type argType, int paramIndex)`

Reply:
Can we rename this function as the `inner` in its name sounded like it was matching the arg and param types of each element of an array?

Reply:
yeah, I'll think of something better.